UV PrimitiveVariable with explicit indices

andrewkaufman opened the pull request:
Prior to this PR, the Cortex convention was to represent the primary UV Set as separate FaceVarying FloatVectorData named "s" and "t". Optionally, "stIndices" represented connectivity information (generally exported from Maya) which could be reverse engineered to produce indexed data (i.e. Vertex "s" and "t" with FaceVaring "stIndices"). Secondary UV Sets followed the same data layout, but were named "other_s", "other_t", and "otherIndices".

This was at odds with most DCCs and Renderers, which tend to prefer float[2] "uv" data. Some, like Maya and Arnold, essentially require the indices as well.

The new Cortex convention for the primary UV Set is a single FaceVarying PrimitiveVariable named "uv". It should contain V2fVectorData for the UVs and can optionally contain FaceVarying IntVectorData indices. If there are indices, the V2fVectorData can be of any size, provided the largest index is within the array. Similarly, secondary UV Sets follow suit, but are just named "other".

To accomplish this, PrimitiveVariable itself has a new member `PrimitiveVariable::indices` and `Primitive` has been updated to account for indices.

Most existing MeshAlgos have been updated for the change, and several Ops have been converted to MeshAlgo and removed. I only converted the Ops that needed updating for the s,t -> uv change.

I've provided on-load conversion for legacy cob/scc files, and updated all the DCC and Renderer code to make the necessary conversions (except those listed below).

There are a few outstanding issues, but I wanted to get this up for review before @johnhaddon goes on holidays. I'll try to outline where we stand right now.

These are the unacceptable issues that should prevent merging:

- IECoreHoudini is not working yet. 653dcc2 is just a WIP commit that breaks several tests. I've emailed SideFx for some advice on how to proceed.

These are the issues I'd be quite happy to ignore until a future PR, if people are amenable to that:

- All the resamplePrimitiveVariable Algos need updating to account for indexed PrimitiveVariables.
- MeshMergeOp should be ported to MeshAlgo, and updated to account for indexed PrimitiveVariables.
- IECoreMaya has 2 failing tests, both todo with MeshMergeOp. I'm tempted to leave them failing, as we know why its happening.
- IECoreRI::GXEvaluator is broken. I haven't attempted to update it for UVs, and I _think_ it'll be going away in the near future, so I'm tempted to just leave it that way...
- IECoreAppleseed doesn't handle UV indices. Note that it didn't before either.
- IECoreNuke FromNukeMeshConverter loses the indices. Note that it did before as well.

johnhaddon commented on src/IECore/MeshAlgoDistributePoints.cpp:
```diff
@@ -0,0 +1,224 @@
+		throw InvalidArgumentException( "MeshAlgo::distributePoints : The input mesh could not be triangulated" );
+	}
+
+	if( result->variables.find( "faceArea" ) == result->variables.end() )
```
Is this use-pre-existing-primvar-if-available behaviour new? I think the Op used to always calculate it, right, rather than trust that the variable means what it thinks it means? I think that generally MeshAlgo functions should be told pretty explicitly what data to operate on rather than make assumptions about naming and validity of data. Other than the obvious P, N etc, we're mostly treating primvars as just being along for the ride, and it being up to the user in Gaffer or wherever to track their meaning/relevance. So here I reckon that'd mean always calculating the areas, or optionally allowing them to be passed in if that's an important optimisation for some known scenarios.

Maybe there's a bigger conversation to be had about general conventions for MeshAlgo, and how it's wrapped in Gaffer? @donboie, any thoughts?

andrewkaufman replied:
Yep, wasn't sure on this one. I made it an optional check to avoid redundant work if, for example, a Gaffer user piped a FaceArea node followed by a Seeds node.

I did pretty intentionally remove the faceAreaPrimVarName options when porting the op though. All those options are just confusion. Nobody ever renames that stuff.

johnhaddon replied:
> Yep, wasn't sure on this one. I made it an optional check to avoid redundant work if, for example, a Gaffer user piped a FaceArea node followed by a Seeds node.

I don't think primvars can be relied on in that way - they're totally arbitrary, and it's down to the user to track their meaning. The mesh could have been deformed since it was calculated, for instance.

> I did pretty intentionally remove the faceAreaPrimVarName options when porting the op though. All those options are just confusion. Nobody ever renames that stuff.

Not sure I follow here. The names don't matter when porting to MeshAlgo, because we're returning PrimitiveVariables rather than adding them to a mesh. In a hypothetical FaceArea node in Gaffer, I would definitely expect there to be a plug to specify the name of the primvar(s) to create, even if it was most often left at the default value. We talked this through when designing the MeshTangents node - the rationale is as follows :

- The presence of the primvar name plugs and their values makes the node more self documenting - "these input variables will be used to generate these output variables".
- The primvar name plugs also allow users more flexibility if they need it - for instance, to generate face areas for two different uv sets without juggling a whole load of intermediate RenamePrimitiveVariable nodes.
- The default values encourage a convention, which makes things easier to understand for users

I definitely want this nailed down as standard when it comes to designing nodes. I'm less concerned about the signatures of the MeshAlgo functions themselves, but I do think we should discuss the merits of forcing the error handling for missing variables onto the caller - that would currently be my preference. @donboie?

andrewkaufman replied:
So should I move this back to forcing re-calculation of face areas on the grounds the user might have deformed the mesh and invalidated those?

johnhaddon replied:
Yes, I definitely think so - that matches the previous behaviour and our current assumptions about the genericity of primitive variables.

It might be interesting to think about mechanisms for storing additional information with some specific primitive variables, so they can be relied upon to have specific semantics. For instance, we could store the hash of the data used to generate them, and then use that to know if they a) represent what we think they do and b) are up to date. I think that's outside the scope of this PR though - maybe something to prototype in Grizzly?

andrewkaufman replied:
Addressed in b839c9ff93e34867fb492cfbcbb1f76a69ac1f48

johnhaddon commented on src/IECore/TriangulateOp.cpp:
```diff
@@ -302,29 +302,34 @@ struct TriangulateOp::TriangulateFn
 
-				it->second.data = data;
+			size_t primVarSize = despatchTypedData<TriangleDataRemap, TypeTraits::IsVectorTypedData>( result.get(), *remap );
+			(void)primVarSize;
```
Seems odd - if we don't need `primVarSize` then we don't need to create the variable at all do we? I think this was necessary before because `primVarSize` was being used in an assertion, right, but that's now removed?

andrewkaufman replied:
Yeah, I had no idea why that was there so I just left it...

andrewkaufman replied:
Addressed in 0a29856fc7a87ac11080b2bca558d5e71854b237

johnhaddon commented on contrib/IECoreAlembic/src/IECoreAlembic/MeshWriter.cpp:
```diff
@@ -152,31 +152,32 @@ class MeshWriter : public PrimitiveWriter
+					if( uvIt->second.indices )
 					{
-						st.push_back( V2f( *sIt, *tIt ) );
+						const std::vector<int> &indexValues = uvIt->second.indices->readable();
```
Hmm, I wonder, should we be storing the indices as unsigned anyway? What's the most common approach? If we do need the conversion, maybe `indexValues( indices.begin(), indices.end() )` would be a more concise way of constructing it?

andrewkaufman replied:
Yeah, I stumbled when I hit this one. But we don't store vertexIds and unsigned, so I think either we change both (in another PR) or we leave it.

andrewkaufman replied:
The trouble with constructing from iterators is that the indices array needs to remain alive until `writeArbGeomParams` has been called or I get hard crashes. That's why I lifted its construction outside the conditional. I guess I could use `std::copy`? I think I tried that and got crashes as well, but that may have been before I realized the lifetime thing...

johnhaddon replied:
> Yeah, I stumbled when I hit this one. But we don't store vertexIds as unsigned, so I think either we change both (in another PR) or we leave it.

Yeah, that's RenderMan legacy I think - that defines everything with ints even if its gotta be positive, and quite a lot of things have followed suit. If we convert indices to unsigned then we'll have fewer conversions to do, but if we make vertexIds unsigned I think we'll be doing more conversions probably. Just wanted to bring it up, but we don't have to deal with it now.

> The trouble with constructing from iterators is that the indices array needs to remain alive until writeArbGeomParams has been called or I get hard crashes. That's why I lifted its construction outside the conditional. I guess I could use std::copy?

Ah, right. Yep, I think that's caught me out before as well. Could just do `indices = vector<unsigned int>( indexValues.begin(), indexValues.end() )`? I _think_ that'll do a move assignment because of C++11 fanciness, although I'd want to check. Or, more old school, `indices.swap( vector<unsigned int>( indexValues.begin(), indexValues.end() ) )`?

andrewkaufman replied:
Addressed in 8843fc72ae5d1bca461a54f5c65e55bebd86491b

johnhaddon commented on contrib/IECoreArnold/src/IECoreArnold/MeshAlgo.cpp:
```diff
@@ -84,121 +84,44 @@ const T *variableData( const PrimitiveVariableMap &variables, const std::string
-
-// Version of the above for when we have no indices available.
-void convertUVSet( const std::string &setName, PrimitiveVariableMap &variables, const vector<int> &vertexIds, AtNode *node )
+void convertUVSet( PrimitiveVariableMap::const_iterator uvIt, const vector<int> &vertexIds, AtNode *node )
```
Iterator seems a strange choice of type, since we don't do any iteration - perhaps `const PrimitiveVariable &`? Ah, I see, we need the name as well as the PrimitiveVariable - still feels a bit odd though...

andrewkaufman replied:
Its anonymous namespace, so I figured the signature wasn't too important to cleanup. I could pass name and variable if you prefer, but it does seem a bit redundant. Also, I'm calling `erase` at the end, which I guess could be lifted out of the function back to the call site.

johnhaddon replied:
> Its anonymous namespace, so I figured the signature wasn't too important to cleanup.

Yeah, not the most important thing for sure. I just felt that passing an iterator implies doing something that needs the iterator, which isn't the case here.

> I could pass name and variable if you prefer, but it does seem a bit redundant.

Or a `const PrimitiveVariableMap::value_type &`, which is what the iterator points to. I do think `name, value` is more readable than `first, second` though.

> Also, I'm calling erase at the end, which I guess could be lifted out of the function back to the call site.

I'm not seeing that - you'd need the container to call erase, and it's not being passed as an argument. The erase appears to already be lifted out into `convertCommon()`. This confusion seems like the best argument for changing the type - there can't be any confusion about the responsibility if `convertUVSet()` has no iterator...

andrewkaufman replied:
Addressed in 214ec0781bc75d1f48bd413abcd407f9cb56e69c

johnhaddon commented on contrib/IECoreArnold/src/IECoreArnold/MeshAlgo.cpp:
```diff
@@ -84,121 +84,44 @@ const T *variableData( const PrimitiveVariableMap &variables, const std::string
+		indicesArray = AiArrayAllocate( indices.size(), 1, AI_TYPE_UINT );
+		for( size_t i = 0, e = indices.size(); i < e; ++i )
+		{
+			AiArraySetUInt( indicesArray, i, indices[i] );
```
Seems to be another vote for unsigned int for indices?

andrewkaufman replied:
Pretty sure this code would be just as ugly though, unless there's some bulk AiArrayUInt thingy?

johnhaddon replied:
Yeah, there's an `AiArrayConvert()` we could use.

johnhaddon commented on contrib/IECoreArnold/src/IECoreArnold/MeshAlgo.cpp:
```diff
@@ -265,37 +185,25 @@ AtNode *convertCommon( const IECore::MeshPrimitive *mesh )
-	uvSetNames.clear();
+	// Find all UV sets. We must perform the iteration to find the
+	// names separately to the iteration to convert them, because
+	// convertUVSet() removes items from variablesToConvert, and
```
convertUVSet doesn't remove items any more...

andrewkaufman replied:
Oops.. you're right. Ignore my comment above where I claimed that it did...

andrewkaufman replied:
Pretty sure I did try condensing that down to one loop and still had issues though.

johnhaddon replied:
I think it can be condensed provided you don't use `it` after `erase( it )`. I think the C++11 idiom is this, because erase now returns the next iterator, whereas before it didn't return anything :

```
for( auto it = c.begin(), it != c(); /* note we don't increment here */ )
{
    if( wantToErase( it ) )
    {
       it = c.erase( it );
    }
    else
    {
        ++it;
    }
}
```

andrewkaufman replied:
Addressed in 214ec0781bc75d1f48bd413abcd407f9cb56e69c

johnhaddon commented on contrib/IECoreArnold/src/IECoreArnold/MeshAlgo.cpp:
```diff
@@ -265,37 +185,25 @@ AtNode *convertCommon( const IECore::MeshPrimitive *mesh )
 	for( PrimitiveVariableMap::iterator it = variablesToConvert.begin(), eIt = variablesToConvert.end(); it != eIt; ++it )
 	{
-		if( it->first == "s" || boost::ends_with( it->first, "_s" ) )
+		/// \todo: add a role enum to PrimitiveVariable, so we can distinguish between UVs and
```
I think a role _and_ a GeometricInterpretation is probably a bit much. I'm still keen on just moving GeometricInterpretation to PrimitiveVariable, and adding a UV enum value to it. I think this will be much simpler.

andrewkaufman replied:
Oh, that's what I intended anyways. Sorry the comment was unclear.

andrewkaufman replied:
Should I maybe just push a new value on the GeometricInterpretation enum now and address these todos that I've scattered across the renderer backends and dcc converters? Then in a future PR we can address moving it to PrimitiveVariable and whatnot.

johnhaddon replied:
Up to you - this PR is already huuge though, so I'd be quite happy to just leave this as a todo rather than introduce any further changes...

andrewkaufman replied:
Addressed in 60097586082178017c5b1ec5fa18b9d5a73eb485 and 8e83cdd0701ae499546cf3f1512a86ab4a6e5908

johnhaddon commented on src/IECoreRI/RendererImplementation.cpp:
```diff
@@ -1828,44 +1828,72 @@ void IECoreRI::RendererImplementation::emitCurvesPrimitive( const IECore::Curves
+// As of Cortex 10, we take a UDIM centric approach
+// to UVs, which clashes with RenderMan, so we must
+// flip the t values before sending to the renderer.
+// We also represent UVs as V2fVectorData and must
```
RenderMan should accept `float[2]` for st I think.

andrewkaufman replied:
What??? Really? So we've been living a lie all these years?

johnhaddon replied:
Yeah. It's my fault.

andrewkaufman replied:
Addressed in 2b7bdbd6460120fb3077ed8929f5e1fa009460b4

andrewkaufman commented:
I can't for the life of me see the error in the Travis log... the tests do pass here though. Any idea what's gone wrong?

Also, bad news on the Houdini front... while they do have a notion of indexed string data (9ec38e7), they don't have a notion of indexed numeric data, so we're going to be stuck expanding uv to FaceVarying (V3f :confused:) and having a side-car uvIndices attrib in Houdini I think. Then we'll have to recontruct the compact version on export. FX TDs the world over will continue to hate us for having to manage uvIndices manually, rather than blaming Houdini itself.

johnhaddon commented:
> All the resamplePrimitiveVariable Algos need updating to account for indexed PrimitiveVariables.

Seems like this might be an important one?

> MeshMergeOp should be ported to MeshAlgo, and updated to account for indexed PrimitiveVariables.

Still important, but less so?

> IECoreMaya has 2 failing tests, both todo with MeshMergeOp. I'm tempted to leave them failing, as we know why its happening.

I think I missed this - why are they failing?

> IECoreRI::GXEvaluator is broken. I haven't attempted to update it for UVs, and I think it'll be going away in the near future, so I'm tempted to just leave it that way...

Yeah, maintaining IECoreRI/GafferRenderMan is becoming a pretty big burden, when what we really want is IECoreDelight with a nice new NSI backend. I'm not too happy about leaving failing tests around though - maybe we rip out GXEvaluator as a precursor to ripping out the old RenderMan backend?

> IECoreAppleseed doesn't handle UV indices. Note that it didn't before either.

Fine.

> IECoreNuke FromNukeMeshConverter loses the indices. Note that it did before as well.

Fine.

andrewkaufman commented:
> I have to also confess to not having anticipated the attendant st->uv renaming across the board.... I guess the argument is that everyone (even me) says UV in practice, so we might as well stop confusing them with another name?

Yes. Unquestionably. I'm soo tired of explaining to users that st means uv. They're called uv literally everywhere other than RenderMan. To be honest, I think non-RenderMan savvy folks think of st as the underlying geometry parameterisation as well, so its a double win.

andrewkaufman commented on src/IECore/Primitive.cpp:
```diff
@@ -391,69 +411,88 @@ void Primitive::hash( MurmurHash &h ) const
 	}
 
-	if( pv.interpolation==PrimitiveVariable::Constant )
+	if( !pv.indices && pv.interpolation == PrimitiveVariable::Constant )
```
@danieldresser-ie made an argument for allowing them provided its a single index 0. The indexed-string stuff in Houdini, for example, would quite naturally produce that variable.

johnhaddon replied:
OK, I see. Is it potentially more complex than allowing a single index of 0 though? For instance, would a Constant array of strings potentially allow any number of indices?

andrewkaufman replied:
Wait, I explained this wrong. Any number of strings. Single index in the range [0,numStrings)

johnhaddon replied:
Wouldn't it be any number of strings and any number of indices in the range [0, numStrings)? To specify a compressed array of strings?

andrewkaufman replied:
Addressed in 32e9845e4deca3a41c0bee3047afdfce40baf9f6

andrewkaufman commented on src/IECore/Primitive.cpp:
```diff
@@ -112,31 +112,167 @@ void Primitive::save( IECore::Object::SaveContext *context ) const
+				bool inserted;
+				PrimitiveVariableMap::iterator uvIt;
+				PrimitiveVariable uvVariable = PrimitiveVariable( uIt->second.interpolation, combineUVs( u, v ) );
+				std::tie( uvIt, inserted ) = variables.emplace( uvSet, uvVariable );
```
I need the iterator afterwards.

andrewkaufman replied:
FYI c++17 will give me a fancy `assign_or_insert` and it can become a one-liner (if we ever get to c++17).

johnhaddon replied:
> I need the iterator afterwards.

Hmm - I'm not seeing that.

andrewkaufman replied:
I think I was doing it at one point... in any case, addressed in 8e7d297e9aab43cd36509750893895d87d15b436

andrewkaufman commented on test/IECore/PrimitiveVariableTest.py:
```diff
@@ -44,59 +44,106 @@ def testConstructors( self ) :
+
 	def testEquality( self ) :
-	
+
```
:nose: goes

johnhaddon replied:
I am not internet savvy enough to understand

andrewkaufman replied:
nose goes is used for tasks that nobody wants to do. last person to put a finger on their nose has to do the task. :middle_finger: :pig_nose:

johnhaddon replied:
Seems I would most definitely be the last person with their finger on their nose. I'll do this at some point when there aren't huge un-merged PRs knocking about.

andrewkaufman commented on test/IECore/PrimitiveTest.py:
```diff
@@ -57,18 +57,107 @@ def test( self ) :
+		self.assert_( not m.isPrimitiveVariableValid( PrimitiveVariable( PrimitiveVariable.Interpolation.Uniform, FloatData( 1 ), IntVectorData( [ 0 ] ) ) ) )
+
+		# constant needs to be vector data if there are indices
+		self.assert_( m.isPrimitiveVariableValid( PrimitiveVariable( PrimitiveVariable.Interpolation.Constant, FloatVectorData( [ 1, 2, 3 ] ), IntVectorData( [ 0 ] ) ) ) )
```
I said above, but the use case is Houdini, which easily generates them (though the data would have been StringVectorData).

andrewkaufman commented on test/IECore/MeshAlgoTest.py:
```diff
@@ -32,478 +32,12 @@
-
-		self.assertEqual( meshReversed["uv"].data, mesh["uv"].data )
-		self.assertEqual( list( meshReversed["uv"].indices ), list( reversed( mesh["uv"].indices ) ) )
+from MeshAlgoDeleteFacesTest import MeshAlgoDeleteFacesTest
```
To be honest, I just did it this way because it made it easy to run all the MeshAlgo tests at once, without the rest of it.

`scons testCorePython TEST_CORE_SCRIPT=test/IECore/MeshAlgoTest.py`

johnhaddon replied:
OK. No biggie. Let's leave as-is until such a time as we make an IECoreTest module, then we can reconsider...

andrewkaufman commented on test/IECore/MeshAlgoDeleteFacesTest.py:
```diff
@@ -35,43 +35,42 @@
 import unittest
 
 import IECore
-from IECore import *
```
For sure. I waffled between having the energy to take it on, and just wanting to move onto the next todo....

johnhaddon replied:
Oh for sure - no need to do it in this PR. Just a general comment...

johnhaddon commented:
> I'm soo tired of explaining to users that st means uv. They're called uv literally everywhere other than RenderMan.

OK, I buy it.

andrewkaufman commented on include/IECore/MeshAlgo.h:
```diff
@@ -47,11 +47,13 @@ namespace MeshAlgo
+PrimitiveVariable calculateFaceArea( const MeshPrimitive *mesh, const std::string &position = "P" );
+
+/// Calculate the face texture area of a mesh primitive based on the specified UV set.
+PrimitiveVariable calculateFaceTextureArea( const MeshPrimitive *mesh, const std::string &uvSet = "uv", const std::string &position = "P" );
```
I did it this way because it seemed simpler to pass names, especially from python, and because `calculateTangents` was already doing it that way. For all these `calculate` algos, it seems to be implied that the variable is part of the primitive, otherwise the calculation wouldn't make sense (i.e why would you calculate texture area of a primitive using texture coordinates from a different primiitve?).

For `resamplePrimitiveVariable()` its explicitly about the variable rather than the primitive, and the resampling is likely in prep for adding to the primitive (rather than adding invalid data and then resampling in place), so that one makes sense to me.

To be honest, I found it a bit bizarre that `deleteFaces()` takes a `PrimitiveVariable` when all it needs is a VectorData. Someone who paints a deletion map is going to have to construct a `PrimitiveVariable` for no reason. Presumably that was to avoid templating or despatching on the client side?

@donboie would you like to weigh in for the new methods?

FYI, the FaceArea node in Gaffer will need to wrap both methods. They are independent calculations.

johnhaddon replied:
> I did it this way because it seemed simpler to pass names, especially from python, and because calculateTangents was already doing it that way.

From Python it's super easy to pass primitive variables - `mesh["name"]`. And like I said, `calculateTangents()` was taking names for a very specific reason.

> For all these calculate algos, it seems to be implied that the variable is part of the primitive, otherwise the calculation wouldn't make sense (i.e why would you calculate texture area of a primitive using texture coordinates from a different primiitve?).

Agreed - that's unlikely. But that's not what I'm trying to get at - I'm talking about making the interface so that the client code is forced to consider what to do if variables are missing. Otherwise if the client code wants to be tolerant of errors, it's either a) doing it's own lookup first anyway, defeating the point of making the algo do it, or b) catching exceptions blindly and hoping they weren't about something else. Hence my question about what we think the behaviour of Gaffer nodes should be if requested variables are missing.

> To be honest, I found it a bit bizarre that deleteFaces() takes a PrimitiveVariable when all it needs is a VectorData.

True. It could take a DataPtr.

andrewkaufman replied:
Just discussing with @donboie over here. So far the best argument for me is by using string names with sensible defaults, all those calculate algo callsites are quite simple for the most common cases:

`result = MeshAlgo::functionName( mesh.get() )`

If we made them take PrimitiveVariables, then all the callsites need to be complex (e.g. no defaults).

`result = MeshAlgo::functionName( mesh.get(), uvPrimVar, pRefPrimVar, pPrimVar )`

The other point we made was maybe this doesn't need to be finalized in this massive PR. We could merge as is, and make another PR prior to an official Cortex 10 release that unifies them all.

johnhaddon replied:
> So far the best argument for me is by using string names with sensible defaults, all those calculate algo callsites are quite simple for the most common cases.

Yeah, that's true. It's definitely nice when the caller knows that stuff _must_ be there, and an exception is the appropriate response to it not being there.

My main concern is Gaffer though, where the caller doesn't know what's coming from upstream, and we need to define what should happen when stuff is missing. If it should be a hard error in Gaffer, then the current interface of passing names and throwing is great. If it should be a silent no-op, the current interface is poor - the caller will need to duplicate all the checks that are currently inside the function, or unconditionally catch all exceptions, potentially silencing something unrelated but important. Neither of you have commented on that.

So, what do you think the appropriate behaviour is in Gaffer, as far as a user is concerned? I can see arguments both ways. Personally I want to see the error right away so I can correct it. But I imagine some users would prefer a more error-tolerant approach, where they can filter a node over a whole range of locations, and have it muddle through as best it can with what's there.

> The other point we made was maybe this doesn't need to be finalized in this massive PR. 

Agreed.

andrewkaufman replied:
> So, what do you think the appropriate behaviour is in Gaffer, as far as a user is concerned? I can see arguments both ways. Personally I want to see the error right away so I can correct it. But I imagine some users would prefer a more error-tolerant approach, where they can filter a node over a whole range of locations, and have it muddle through as best it can with what's there.

Coming from a Houdini background, I want both in Gaffer, by introducing a Warning state (to complement the Error state we have now). If my graph is:

`SceneReader -> FaceArea (filtered) -> OSLObject (filtered) -> Seeds (filtered)`

If I'm filtering a single object explicitly, I'd want the Error (red icon) on the FaceArea node when "uv" is missing. If I was filtering a bunch of objects, I'd probably want my stream to continue un-interrupted, but also non-silently (hence the per-object warning in yellow).

In either case, it seems the exception string thrown by the algo is likely to be the most useful bit of information. The node might still catch and re-raise to append some useful bits:

```
Warning : Invalid mesh at "/location/of/my/bad/mesh". MeshAlgo::distributePoints : MeshPrimitive has no suitable "uv" primitive variable.
```

We end up with some redundant bits maybe, but that would be an argument for an IECore::Message (to seperate out the prefix) rather than an exception I think.

johnhaddon replied:
I see - that makes sense I guess. I guess you also want this warning status to propagate around to the downstream nodes and any parent nodes? That would require a bit of thought - we currently do it using exception propagation, but we won't be able to do that for warnings. I also don't know how we'll figure out when to make it an error and when to make it a warning - introspecting the filter doesn't seem right given that it's potentially very expensive to go looking for other matches in the scene.

andrewkaufman replied:
I don't know if they'd need to propogate as errors do... maybe they only show when you're explicitly inspecting that node in the graph?

andrewkaufman replied:
in houdini, when to error vs warn is a per-node plug. not sure if that's sensible for gaffer, but in the above network I might want to set my Seeds to error and my OSLObject to warn.

andrewkaufman commented:
>> All the resamplePrimitiveVariable Algos need updating to account for indexed PrimitiveVariables.
> Seems like this might be an important one?

Yeah... and a f'ing lot of code to change and fiddly to test. @donboie mentioned that maybe we could do the dumb `expandedVariableData` thing for now, so it just works, and then fix it up to maintain indices in a future PR. I guess that's where the `PrimitiveVariable` not necessarily being part of the `Primitive` will screw us though...

andrewkaufman commented:
>> MeshMergeOp should be ported to MeshAlgo, and updated to account for indexed PrimitiveVariables.

> Still important, but less so?

Yep, but also quite a bit of redesign, since I assume it'll need to be `MeshPrimitivePtr MeshAlgo::mergeMeshes( const std::vector<const MeshPrimitive *> &meshes )` rather than just mesh3 = mesh1 + mesh2.

andrewkaufman commented:
>> IECoreMaya has 2 failing tests, both todo with MeshMergeOp. I'm tempted to leave them failing, as we know why its happening.

> I think I missed this - why are they failing?

Because indexed UVs are being handed to MeshMergeOp and the result is not as it should be.

andrewkaufman commented:
>> IECoreRI::GXEvaluator is broken.

> I'm not too happy about leaving failing tests around though - maybe we rip out GXEvaluator as a precursor to ripping out the old RenderMan backend?

Sure.

johnhaddon commented:
>>> All the resamplePrimitiveVariable Algos need updating to account for indexed PrimitiveVariables.

>> Seems like this might be an important one?

> Yeah... and a f'ing lot of code to change and fiddly to test. @donboie mentioned that maybe we could do the dumb `expandedVariableData` thing for now, so it just works, and then fix it up to maintain indices in a future PR. I guess that's where the `PrimitiveVariable` not necessarily being part of the `Primitive` will screw us though...

Off the top of my head, for upsampling we'd just resample the indices, and it's only for downsampling we'd use the expanded data? In the long run we might want to do that with the fancy indexing_iterator thing we talked about, but for now using explicitly expanded data seems fine.

Seems like we should add a `PrimitiveVariable::expandedData()` method and then call that from  `Primitive::expandedVariableData()`?

andrewkaufman commented:
I've pushed changes which I believe address all the review comments. I'll go mark each one to make sure.

I've also added GeometricData::UV, which we were originally planning for a separate PR, but it seemed easy enough todo now, and helped address one of the review comments.

Note that IECoreHoudini is still not working yet, so we can't merge. I've dropped the original bogus commit for that, and I'll push a new one when I get something workable.

andrewkaufman commented on include/IECore/PrimitiveVariable.h:
```diff
@@ -80,6 +82,12 @@ struct IECORE_API PrimitiveVariable {
+	/// Optional indices array into the primary Data. This can be used
+	/// to store a more compact representation of the data, for example
+	/// a constant list of strings which are mapped to Vertex, Uniform,
+	/// or FaceVarying via the indices. Similarly, UV sets can be stored
```
Addressed in 0e48204537c7a17067cba40aa67990549ff909fa

andrewkaufman commented on src/IECore/PrimitiveVariable.cpp:
```diff
@@ -72,11 +80,22 @@ bool PrimitiveVariable::operator==( const PrimitiveVariable &other ) const
+		indicesEqual = indices->isEqualTo( other.indices.get() );
+	}
+
+	return dataEqual && indicesEqual;
```
Addressed in 526749152d1dbe5a6166f6ecc3eddb908b0b3e32

andrewkaufman commented on include/IECore/Primitive.inl:
```diff
@@ -59,15 +64,53 @@ const T *Primitive::variableData( const std::string &name, PrimitiveVariable::In
+	T *data = runTimeCast<T>( it->second.data.get() );
+	if( !it->second.indices )
+	{
+		return data->copy();
```
Addressed in 69d32e769507b2b61d8fecb26defcb2aaf77d75c

andrewkaufman commented:
> All the resamplePrimitiveVariable Algos need updating to account for indexed PrimitiveVariables.

Addressed in 2a7473b7fc7e244ca943118514d9de7842961d5a, 8b3f6ce02a14bb5cd4a6bd61458646dc09019fa8, and eaad42aac607e1969fb7c4716cc973a467de12fe

> MeshMergeOp should be ported to MeshAlgo, and updated to account for indexed PrimitiveVariables.

Update to account for PrimitiveVariables addressed in de301f7e9edd7908c76bdc052908351d1b236eb3. Port to MeshAlgo left for a future PR.

> IECoreMaya has 2 failing tests, both todo with MeshMergeOp.

Addressed in de301f7e9edd7908c76bdc052908351d1b236eb3

> IECoreRI::GXEvaluator is broken. I haven't attempted to update it for UVs, and I think it'll be going away in the near future, so I'm tempted to just leave it that way...

Addressed in 28ac5ccb5430aae15b07df6bc88b6b5cc2ee2c53

andrewkaufman commented:
Ok, this should be ready for final review now. For the Houdini stuff, I decided to force UV index expansion going into Houdini, and detect duplicate to regenerate indices coming out.

@donboie can you take a look since @johnhaddon is still away for a few days?

andrewkaufman commented:
Also added a commit to set `GeometricData::UV` when loading Alembics, since I forgot that one on the last push.

andrewkaufman commented:
Note that Travis still had those odd failures that people have seen on other PRs, but the tests are passing at IE...

johnhaddon commented on src/IECoreRI/NSI/AttributeAlgo.cpp:
```diff
@@ -50,6 +50,7 @@ NSIType_t convertGeometricInterpretation( GeometricData::Interpretation interpre
 	switch( interpretation )
 	{
 		case GeometricData::None :
+		case GeometricData::UV :
```
This isn't right, because NSITypeVector implies 3 floats, whereas UV implies 2. In practice it won't matter because this function isn't called for non-V3f data, but just returning NSITypeInvalid for UV might make any future problems easy to diagnose...

andrewkaufman replied:
Addressed in 5bec0cf799fa89ba7a3bf27c33c3047924185653

andrewkaufman commented:
Any other notes on this one @johnhaddon ? Or just wondering about the Travis debacle at this point?

johnhaddon commented:
> Any other notes on this one @johnhaddon?

Nope - I'm just getting Gaffer updated to work with this as a final validation. I would like to take a last look though the resampling code tomorrow, but if @donboie already gave it a once over then I'll leave that - I think he knows that code better than me anyway.

> Or just wondering about the Travis debacle at this point?

Would be good to get this sorted. I wonder if dropping the "-j 2" argument to scons might be instructive? Maybe SCons is doing things in parallel when it shouldn't because we've failed to declare a dependency somewhere?

andrewkaufman commented:
> I wonder if dropping the "-j 2" argument to scons might be instructive? 

```
Retrieved `test/IECore/IECoreTest' from cache
test/IECore/IECoreTest > test/IECore/results.txt 2>&1
scons: *** [test/IECore/results.txt] Error 201
scons: building terminated because of errors.
The command "scons testCore testCorePython CXX=$CXX CXXSTD=$CXXSTD DEBUG=$DEBUG ENV_VARS_TO_IMPORT="PATH TRAVIS LD_PRELOAD" RMAN_ROOT=$DELIGHT BOOST_LIB_SUFFIX="" BUILD_CACHEDIR=sconsCache" exited with 2.
```

Do you think the ` test/IECore/results.txt` exists in this case? Would it have more info in there? Maybe I can get it to echo the contents of that and config.log ?

andrewkaufman commented:
> I would like to take a last look though the resampling code tomorrow, but if @donboie already gave it a once over then I'll leave that

FWIW, we have an internal ticket for @donboie to refactor that some more after this gets merged.

johnhaddon commented:
> Do you think the test/IECore/results.txt exists in this case? Would it have more info in there? Maybe I can get it to echo the contents of that and config.log ?

Sounds worth a try.

andrewkaufman commented:
Seems like a real failure (though I'm still unable to reproduce here). Probably worth keeping my cat command for the future I guess...

```
test/IECore/ComputationCacheTest.cpp(149): error in "ComputationCacheTest::test": check *weirdValue == *cache.get( ComputationParams(1), Cache::NullIfMissing ) failed
```

johnhaddon commented:
While testing in Gaffer, I ran across problems trying to texture a mesh with a GLSL shader in the viewport. [This commit](https://github.com/johnhaddon/cortex/commit/79d43a10a392ba8b18639e5efcfe5d8bbd3689bc) fixes IECoreGL to use the new "uv" naming in place of the old "st", and gets things working again.

johnhaddon commented:
One more commit I need in order to get Gaffer working with this - [support for V2fVectorData in OpenImageIOAlgo::DataView](https://github.com/johnhaddon/cortex/commit/bf826a273b1df56bc77734db90b5fdb15c63cecd). Could you cherry-pick both commits onto your branch please?

johnhaddon commented:
Thanks - I'll have one more cherry for you shortly. There's a bug in `resamplePrimitiveVariable` that showed up in the OSLObject tests. I have a fix but haven't yet written the Cortex unit tests to prove it...

andrewkaufman commented:
Is it todo with uninitialized memory by any chance? If so, @donboie might have the Cortex unit test to prove it

johnhaddon commented:
It is indeed to do with uninitialised memory.

donboie commented:
I have a  production case which demonstrates the problem, not a test case. I think it's stuff like this template   https://github.com/ImageEngine/cortex/blob/59aa9f28fecea7a59973d2c6ab5f4f1a7e0d8d89/src/IECore/MeshAlgo.cpp#L440 when used with Imath vector types.

johnhaddon commented:
Yep, that's exactly what I've fixed. Could you cherry-pick [this commit](https://github.com/johnhaddon/cortex/commit/e2e0146b9fecc45c74620f9d7f84b36c8dac2811) Andrew? It contains both a simple test case and the fix...

andrewkaufman commented:
Sure thing. Also note that 7e2268915a3f673d13e850812634c7ad5592e703 has the Travis tests passing now, and the tests still pass at IE. No idea why clearing that ObjectPool is required to get Travis working, nor if its invalidating the test?

andrewkaufman commented:
Woohoo. Passing still. Thoughts on that change?

johnhaddon commented:
> Thoughts on that change?

The ComputationCacheTest one? I _think_ it's kosher. The `weirdValue` is the third value to be pushed into that pool, and the pool has size 2. My LRUCache changes meant that the parallel cache went from using a perfect LRU order to using a second chance sweep algorithm, so we can no longer guarantee that weirdValue won't be removed even though it was last in. But why does it pass for you and not on Travis? I think this is because the parallel cache uses bins internally, one bin per core, and the sweep iterations visit those in order. Different core counts would therefore yield a different iteration order and a different object being evicted. Testing locally I've found that a count of 5 leads to failure whereas others are OK.

I've submitted a test job to Travis to find out how many cores are being reported there to verify this hypothesis.

Either way, I don't believe this relates to your work on this branch at all...

andrewkaufman commented:
> Either way, I don't believe this relates to your work on this branch at all...

I don't believe it does either (the failure is on all current PRs targetting master). Does that mean you'd like me to drop it off this branch and let this one have failing travis tests?

johnhaddon commented:
> I don't believe it does either (the failure is on all current PRs targetting master). Does that mean you'd like me to drop it off this branch and let this one have failing travis tests?

Got my tests back from Travis - 2 cores, so 2 bins. And that _passes_ here - infuriating! But, we're testing locally with boost 1.61, and Travis has boost 1.55. The [release notes for boost 1.56](http://www.boost.org/doc/libs/1_62_0/libs/multi_index/doc/release_notes.html#boost_1_56) state that `multi_index::hashed_unique` changed data structures in that version, so it didn't seem implausible that iteration order also changed at that time. So I rebuilt locally using 1.41 (don't have 1.55 to hand) and got an identical failure to the one on Travis. Phew.

I don't mind whether you want to make a separate PR or keep the commit on this branch, but could you update the commit message to properly explain the change and its justification? I'm signing off for the night now, but this all looks good to me, so perhaps you could merge, and then I'll make an equivalent Gaffer PR tomorrow...

Cheers...
John

andrewkaufman commented:
Ok, I just forced a rebase that puts the 2 travis related commits after the cherry-picked commits, and rewords the test change commit to follow your explanations.

johnhaddon commented:
Merged, and with a day to spare! :tada:
